flake.nix

```diff
@@ -19,12 +19,14 @@
         dependencies' = pkgs.lib.lists.concatMap (builtins.match "([^>=<]*).*") pyproject.project.dependencies;
         # the package is called gpg on PyPI but gpgme in nixpkgs
         dependencies = map (x: if x == "gpg" then "gpgme" else x) dependencies';
+        version = "0.11.dev+${if self ? shortRev then self.shortRev else "dirty"}";
```

Comment:
The architecture of nix has two distinct phases: 
1. the evaluation phase where the nix code is evaluated, in order for this to be deterministic we need all (flake-) dependencies in the lock file. The result is a build plan for step 2. 
2. the build phase: this happens in a sandbox without internet access, we can run all the commands and use all the files that we included in our build plan and produce some directory structure and files as a result.

The consequence of this is that we can not run arbitrary commands (e.g. python) during phase 1. We could read static files from disk from all our flake inputs and the git repo itself (because they are considered constant/deterministic). 

The postfix of the version in `${if ... then ... else ...}` is special nix flake related code that produces the short git hash if we build from a clean git state and the string "dirty" otherwise.

Now I also want to add a small prefix to the version string.  If python/pip accepts this I could just use "dev". I have to check https://peps.python.org/pep-0440/ again. Maybe we can also use "0.dev" or something that is obviously "wrong" so people pay more attention to the git hash in the version.
I will have a look at this when I am at home again and come back to you.

PS: Strictly speaking nix does not need the version. As a package manager it does not use version constraints. All dependencies are always specified exact via git hashes or download urls+sha hashes. The version is mainly used for pip to integrate it into the code and to append it to the resulting nix-store path (the location where alot is installed on disk) for humans to see it.

Reply:
thanks for the elaborate response.
Then I suppose we can simply use an obviously false verstion string like "XXX" or similar?